src/utils/ucc_dt_reduce.h

```diff
@@ -0,0 +1,134 @@
+/**
+ * Copyright (C) Mellanox Technologies Ltd. 2021.  ALL RIGHTS RESERVED.
+ * See file LICENSE for terms.
+ */
+
+#ifndef UCC_DT_REDUCE_H_
+#define UCC_DT_REDUCE_H_
+
+#include "components/mc/ucc_mc.h"
+#include "components/ec/ucc_ec.h"
+
+static inline ucc_status_t
+ucc_dt_reduce(void *src1, void *src2, void *dst, size_t count,
+              ucc_datatype_t dt, ucc_memory_type_t mem_type,
+              ucc_coll_args_t *args)
+{
+    if (!UCC_DT_IS_PREDEFINED(dt)) {
```

Comment:
should we flip the order so that the more common use-case (UCC_DT_IS_PREDEFINED) comes in the if path instead of else? Or add ucc_unlikely()? It's a micro optimization that could help the branch prediction.

Reply:
i don't think it is a proper place for unlikely. If the runtime actually uses non-default dtype there will be always incorrect hint. We don't have prior knowledge about it.

Reply:
changed order